### PR TITLE
DR-1043 Add retry to insert queries in the BigQueryPdao

### PR DESCRIPTION
**Pull request by nmalfroy**
This should greatly reduce the rateLimit errors we've been seeing the `snapshotByQueryHappyPathTest` test where we are doing more than 5 updates on a single dataset in a 10 second window.

The default values are to retry 3 times (so try up to 4 times) with a 500ms pause in between tries.  These values can be overridden

Here's an example of where that shows up:
image

### CODE REVIEW

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java [hunk 1]
```diff
@@ -89,10 +92,16 @@
...
+    private static final int DEFAULT_MAX_NUM_UPDATE_LIMIT_INDUCED_RETRIES = 3;
+    private static final int DEFAULT_RETRY_WAIT_MS = 500;
+
     private final String datarepoDnsName;
     private final DataLocationService dataLocationService;
     private final DatasetService datasetService;
 
+    @Value("${datarepo.bq.rateLimitRetries}") private Integer rateLimitRetries;
+    @Value("${datarepo.bq.rateLimitRetryWaitMs}") private Integer rateLimitWait;
+
```
Comment (RoriCremer) [inline src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: why not use the config enum for these guys?

**Review by RoriCremer — APPROVED** [review]

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java (hunk 1 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: Ooh...what's the config enum? It does sound like I thing I should use :)

Comment (ddietterich) [reply src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: We never use $Value. It is considered harmful: harder to find the source of configuration, not well typed, ...
Instead, we use configuration classes that map to specific properties. Then use getters to access those classes.
See `main/java/bio/terra/app/configuration/ApplicationConfiguration` for the main one.

If you want to have run-time configurability of a config item, then you need to do additional work that Rori referred to:
* add an entry to `.../service/configuration/ConfiguEnum.java`
* add a mapping in `.../service/configuration/ConfigurationService.java` toward the bottom of the file.

You will see in some of our tests that we manipulate parameters by hitting a configuration endpoint - the same one used to enable faults.

**Review by ddietterich — CHANGES_REQUESTED** [review]
I tried to leave this comment inline, but github was saying "line must be part of the diff"!?
Anyway, line 1654 of the diff is catching InterruptedException.

We should not be rewriting InterruptedExceptions. If that was in the code before, it should not have been.

We use InterruptedException to cleanly shutdown thread pools when the pod is shutting down. The effect of rewriting the exception is to cause the Stairway flight to fail with that PdaoException rather than restart when the pod comes back.

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java (hunk 1 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: Got it. I'll add a new config object for the BigQuery config.  I'm going to hold off for the runtime config...that seems to be geared towards test parameters.  Also not sure that it's a runtime change.  Seems reasonable to have these effectively tied to the deployment

**Comment by nmalfroy** [comment]
@ddietterich are you saying that the interrupted exception should just throw?  I went by what was in there now.  I'll just have it pass through, I'd just like to know where the IE get's handled

**Comment by ddietterich** [comment]
In general, our code either runs in a Stairway flight or it does not. You will see some cases where services catch InterruptedException because they are using some code that runs in both contexts.

For code that runs in a Stairway flight, the top frame of the flight thread is code that catches the InterruptedException and tries to set state in the Stairway database to help with future recovery. I think there is no code in BigQueryPdao that runs outside of a Stairway flight.

**Comment by nmalfroy** [comment]
Ok, updated with PR comments hopefully addressed

**Review by ddietterich — APPROVED** [review]
lgtm - thanks!

**Comment by nmalfroy** [comment]
Perfect, thanks!

**Review by fboulnois — APPROVED** [review]
This will be fabulous for the alpha tests. lgtm!

**Pull request merged**